spec.html

```diff
@@ -648,28 +651,46 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
         1. For each Module _m_ of _module_.[[AsyncParentModules]], do
-          1. Decrement _m_.[[PendingAsyncDependencies]] by 1.
-          1. If _m_.[[PendingAsyncDependencies]] is 0 and _m_.[[EvaluationError]] is ~empty~, then
+          1. If _execList_ does not contain _m_ and _m_.[[CycleRoot]].[[EvaluationError]] is ~empty~, then
+            1. Assert: _m_.[[EvaluationError]] is ~empty~.
             1. Assert: _m_.[[AsyncEvaluating]] is *true*.
-            1. If _m_.[[CycleRoot]].[[EvaluationError]] is not ~empty~, return *undefined*.
-            1. If _m_.[[Async]] is *true*, then
-              1. Perform ! ExecuteAsyncModule(_m_).
-            1. Otherwise,
-              1. Let _result_ be _m_.ExecuteModule().
-              1. If _result_ is a normal completion,
-                1. Perform ! AsyncModuleExecutionFulfilled(_m_).
-              1. Otherwise,
-                1. Perform ! AsyncModuleExecutionRejected(_m_, _result_.[[Value]]).
+            1. Set _m_.[[PendingAsyncDependencies]] to _m_.[[PendingAsyncDependencies]] - 1.
+            1. If _m_.[[PendingAsyncDependencies]] is equal to 0, then
+                1. Append _m_ to _execList_.
+                1. If _m_.[[Async]] is *false*, perform ! GatherAsyncParentCompletions(_m_, _execList_).
+        1. Return *undefined*.
+      </emu-alg>
+      <emu-note>
+        <p>When an async execution for a root _module_ is fulfilled, this function determines the list of modules which are able to synchronously execute together on this completion, populating them in _execList_.</p>
+      </emu-note>
+    </emu-clause>
+
+    <emu-clause id="sec-async-module-execution-fulfilled" aoid="AsyncModuleExecutionFulfilled">
+      <h1><ins>AsyncModuleExecutionFulfilled ( _module_ )</ins></h1>
+      <emu-alg>
+        1. Let _execList_ be a new empty List.
+        1. Assert: _module_.[[AsyncEvaluating]] is *true*.
+        1. Perform ! GatherAsyncParentCompletions(_module_, _execList_).
+        1. Let _sortedExecList_ be a List of elements that are the elements of _execList_, in the order in which they had their [[AsyncEvaluating]] fields set to *true* in InnerModuleEvaluation.
+        1. Assert: All elements of _sortedExecList_ have their [[AsyncEvaluating]] field set to *true*, [[PendingAsyncDependencies]] field set to 0 and [[EvaluationError]] field set to *undefined*.
+        1. For each Module _m_ of _sortedExecList_, do
+          1. If _m_.[[AsyncEvaluating]] is *false*, then
+            1. Assert: _m_.[[EvaluatingError]] is not ~empty~.
+          1. Otherwise, if _m_.[[Async]] is *true*, then
+            1. Perform ! ExecuteAsyncModule(_m_).
+          1. Otherwise,
+            1. Let _result_ be _m_.ExecuteModule().
+            1. If _result_ is an abrupt completion,
+              1. Perform ! AsyncModuleExecutionRejected(_m_, _result_.[[Value]]).
+            1. Otherwise, set _m_.[[AsyncEvaluating]] to *false*.
```

Comment:
Ok, the bug I mentioned yesterday actually has it's root here.

When we set module to false here, we are doing it before it calls it's own GatherAsyncParentCompletions, and we end up checking for an error which doesnt exist. That throws on the assertion.

We should instead move this flag change to after this loop, and only set it for the current module, because all modules (async or no) will now call asyncModuleExecutionFulfilled.

Reply:
AsyncModuleExecutionFulfilled should only be called for modules that have `[[Async]]` set to true, as we only run `ExecAsyncModule` on these async modules.

The `AsyncModuleExecutionFulfilled` always can only happen after the promise completion of the module body, so we are guaranteed that this routine should finish synchronously before any of the completions trigger.

It is only for the sycn modules that do not go through async exec that we thus set `AsyncEvaluating` to false explicitly here.

Let me know if that clarifies the logic?

Reply:
Oh, i think this changed since I last looked at it. Let me look again

Reply:
Yes, this is fixed in this version! I had the old behavior from before 279739dc10d43acb86c471632a780c3c1d7a1f5f

Reply:
Ah yes, that was actually fixed with your previous feedback in https://github.com/tc39/proposal-top-level-await/pull/159#discussion_r576748351.